// @sap/ux-specification/dist/src/sync/v4/import/app/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V4FreestyleAppProvider = exports.V4AppProvider = exports.getSettings = exports.getPagesAndConfigs = exports.getHome = exports.getPages = void 0;
var baseAppProvider_1 = require("./baseAppProvider");
Object.defineProperty(exports, "getPages", { enumerable: true, get: function () { return baseAppProvider_1.getPages; } });
Object.defineProperty(exports, "getHome", { enumerable: true, get: function () { return baseAppProvider_1.getHome; } });
Object.defineProperty(exports, "getPagesAndConfigs", { enumerable: true, get: function () { return baseAppProvider_1.getPagesAndConfigs; } });
Object.defineProperty(exports, "getSettings", { enumerable: true, get: function () { return baseAppProvider_1.getSettings; } });
var appProvider_1 = require("./appProvider");
Object.defineProperty(exports, "V4AppProvider", { enumerable: true, get: function () { return appProvider_1.V4AppProvider; } });
var freestyleProvider_1 = require("./freestyleProvider");
Object.defineProperty(exports, "V4FreestyleAppProvider", { enumerable: true, get: function () { return freestyleProvider_1.V4FreestyleAppProvider; } });
